### PR TITLE
LPS-110927 Update Liferay.Portlet.openWindow utility

### PR DESCRIPTION
**Pull request by markocikos**
Hey @wincent ,

~~This PR contains LPS-110926 commits, so we cannot forward it. We can safely assume https://github.com/brianchandotcom/liferay-portal/pull/87801 will be merged with no additional changes, so we can proceed with review.~~

In this PR:
- We are adding `Liferay.Portlet.openModal` utility. It is implemented as a wrapper around Clay-based `Liferay.Util.openModal`. It is a direct replacement of AUI-based `Liferay.Portlet.openWindow`.
- We are extending `openModal` API to support `headerHTML`.
- We are implementing support for form buttons within iframe
- We are changing implementation of `Liferay.Portlet.openWindow`, to be a thin wrapper around `Liferay.Portlet.openModal`
- We are replacing all usages of `Liferay.Portlet.openWindow`

To test staging-configuration-web
1. Enable Local Live staging (Control Panel > Site > Publishing > Staging)
1. Go to home page > Staging tab
1. Open home page > Hello World porlet options > Staging

To test export-import-web
1. Open home page > Hello World porlet options > Export / Import

To test notifications-web
1. Click user icon on top right > Notifications
1. Open Notifications porlet options > Configuration

To test portal-web
1. Open home page > Hello World porlet options > Configuration

Notes:
- We could have just changed implementation of `Liferay.Portlet.openWindow`, but I like to have consistent `*.openModal` for Clay-based utils.
- We could have just marked `Liferay.Portlet.openWindow` as deprecated, like we did with `SignInModal`, but we want to aggressively remove eager AUI code, like in `portlet.js`. Among other things, it is a performance bottleneck for AUI preloading and requires special babel configuration.
- We need to implement support for buttons withing iframe, see the reasoning [here](https://github.com/liferay-echo/liferay-portal/pull/237#issuecomment-614559365). While most buttons work by default, we need to have special handling for 'cancel' button. It needs to close modal. To achive this, I added delegation, defined on iframe document node. To be able to do this, I needed iframe `ref`. As `ClayModal.Body` does not expose this, I replaced it with plain `<div>`. All of this causes a bit of code duplication, with CSS classes around body. 
Is there a batter way to this this? Could `ClayModal` expose iframe `ref`, or is that a bad practice? (@kresimircoko this might be interesting for you).
- On the track of https://github.com/wincent/liferay-portal/pull/208#discussion_r411295441, I tweaked body html to implement `url` / `bodyHTML` exclusion, each leading to it's own subcomponent.

Please review the code.

Thank you!

/cc @jbalsas

### CODE REVIEW

**Comment by markocikos** [comment]
Rebased, tested, force-pushed. No changes. LPS-110926 commits have been merged.

File: modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/liferay/global.es.js [hunk 1]
```diff
@@ -96,6 +96,18 @@ Liferay.Portlet = Liferay.Portlet || {};
...
+		commands.openPortletModal(...args);
+	});
+};
+
+Liferay.Portlet.openWindow = (...args) => {
+	Liferay.Loader.require('frontend-js-web/liferay/modal/Modal', commands => {
+		commands.openPortletWindow(...args);
+	});
+};
+
```
Comment (wincent) [inline modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/liferay/global.es.js:110]: Given that we now have multiple `Liferay.Loader.require` calls for this module, I wanted to see what would happen if we ever wound up in a race condition where multiple places called one of these functions over a slow network.

So I picked a module that wasn't yet loaded and loaded it half a dozen times with:

```
{
Liferay.Loader.require('segments-web@2.0.9/js/components/shared/ClayToggle.es', () => console.log(Date.now()));
Liferay.Loader.require('segments-web@2.0.9/js/components/shared/ClayToggle.es', () => console.log(Date.now()));
Liferay.Loader.require('segments-web@2.0.9/js/components/shared/ClayToggle.es', () => console.log(Date.now()));
Liferay.Loader.require('segments-web@2.0.9/js/components/shared/ClayToggle.es', () => console.log(Date.now()));
Liferay.Loader.require('segments-web@2.0.9/js/components/shared/ClayToggle.es', () => console.log(Date.now()));
Liferay.Loader.require('segments-web@2.0.9/js/components/shared/ClayToggle.es', () => console.log(Date.now()));
}
```

and then another 6 after it had loaded. Results:

CleanShot 2020-04-22 at 16 21 37@2x

Conclusions:

- When requests are in still in flight, you can wind up with multiple requests to the `js_resolve_modules` endpoint.
- As soon as one of those `js_resolve_modules` requests finalizes, we fetch the module once.
- Subsequent requests don't call `js_resolve_modules`, or hit the network to fetch the module.

I think the likelihood of a race with these functions is small, so no need to worry about it. But I thought it was interesting question anyway.

File: modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/liferay/modal/Modal.js [hunk 2]
```diff
@@ -104,22 +157,60 @@ const Modal = ({bodyHTML, buttons, id, onClose, size, title, url}) => {
...
+		const src = iframeURL.toString();
+
+		const initCancelButtonDelegation = () => {
+			dom.delegate(
+				iframeRef.current.contentWindow.document,
+				'click',
+				'.btn-cancel',
+				() => {
+					processClose();
+				}
```
Comment (wincent) [inline modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/liferay/modal/Modal.js:200]: Would probably look better as:

```js
() => processClose()
```

File: modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/liferay/modal/Modal.js [hunk 3]
```diff
@@ -104,22 +157,60 @@ const Modal = ({bodyHTML, buttons, id, onClose, size, title, url}) => {
...
+		if (iframeBodyCssClass) {
+			bodyCssClass = `${bodyCssClass} ${iframeBodyCssClass}`;
+		}
+
+		iframeURL.searchParams.set(`_${namespace}_bodyCssClass`, bodyCssClass);
+
+		const src = iframeURL.toString();
+
+		const initCancelButtonDelegation = () => {
+			dom.delegate(
```
Comment (wincent) [inline modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/liferay/modal/Modal.js:194]: I am not super familiar with how `dom.delegate` works. I know it returns a handle though, and perhaps we should be calling `removeListener()` on it when we unmount so as not to leak resources. What do you think?

**Comment by markocikos** [comment]
> What do you think about adding a spinner?

We already have a task for this: https://issues.liferay.com/browse/LPS-111118

See more discussion about it in https://github.com/liferay/clay/issues/3111

There is no blocker on this task, as we both have a workaround for iframe `onLoad` in this thread, as well as `iFrameProps` added in https://github.com/liferay/clay/pull/3139

**Comment by wincent** [comment]
So if understand what's in all those links correctly, seems like the conclusion was:

- `iFrameProps` will be in Clay release soon.
- We can use that to hook into `onLoad`.
- In the meantime, in this PR we [are managing our own iframe manually](https://github.com/wincent/liferay-portal/blob/caf4728976146056a72a4b676ddf5e7141ac1167/modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/liferay/modal/Modal.js).
- At some future time, Clay may handle spinner for us automatically.

But I still don't know what you want to do: Implement temporary spinner in this PR? Pause this PR until we get `iFrameProps`? Until Clay provides auto-spinner? Ship anyway without spinner then fast-follow?

**Comment by markocikos** [comment]
> But I still don't know what you want to do: Implement temporary spinner in this PR? Pause this PR until we get iFrameProps? Until Clay provides auto-spinner? Ship anyway without spinner then fast-follow?

My original plan was to to do a fast-follow, but I guess it makes sense to add spinner to this PR. It is pretty wide reaching.

Clay handling spinner automatically may not be useful for this util, it's likely that we cannot use it because of the `ref` problem. Trying to utilize this can be left for the future.

**Comment by kresimir-coko** [comment]
> Is there a batter way to this this? Could ClayModal expose iframe ref, or is that a bad practice? (@kresimircoko this might be interesting for you).

Just a heads up @markocikos you used my private, old GitHub handle, one not attached to Liferay.

**Comment by markocikos** [comment]
See https://github.com/wincent/liferay-portal/pull/230

**Pull request closed without merging**